kubert/src/lease.rs

```diff
@@ -0,0 +1,469 @@
+            })))
+            .await?;
+
+        let claim = Claim {
+            holder: claimant.to_string(),
+            expiry: now + lease_duration,
+        };
+        let meta = Meta {
+            version: lease
+                .metadata
+                .resource_version
+                .ok_or(Error::MissingResourceVersion)?,
+            transitions: meta.transitions + 1,
+        };
+        Ok((claim.into(), meta))
+    }
+
+    /// Renew the lease (i.e. assuming the claimant IS the current holder of the
+    /// lease).
+    ///
+    /// A strategic merge is used so that only the `renewTime` field is updated
+    /// in most cases. The `leaseDurationSeconds` fields may also be updated if
+    /// the caller passed an updated value.
+    async fn renew(
+        &self,
+        meta: &Meta,
+        claimant: &str,
+        params: &ClaimParams,
+    ) -> Result<(Arc<Claim>, Meta), Error> {
+        let lease_duration = chrono::Duration::from_std(params.lease_duration)
+            .unwrap_or_else(|_| chrono::Duration::max_value());
+        let now = chrono::Utc::now();
+        let lease = self
+            .patch(&kube_client::api::Patch::Strategic(serde_json::json!({
+                "apiVersion": "coordination.k8s.io/v1",
+                "kind": "Lease",
+                "metadata": {
+                    "resourceVersion": meta.version,
+                },
+                "spec": {
+                    "renewTime": metav1::MicroTime(now),
+                    "leaseDurationSeconds": lease_duration.num_seconds(),
+                },
+            })))
+            .await?;
+
+        let claim = Claim {
+            holder: claimant.to_string(),
+            expiry: now + lease_duration,
+        };
+        let meta = Meta {
+            version: lease
+                .metadata
+                .resource_version
+                .ok_or(Error::MissingResourceVersion)?,
+            transitions: meta.transitions,
+        };
+        Ok((claim.into(), meta))
+    }
+
+    async fn patch<P>(
+        &self,
+        patch: &kube_client::api::Patch<P>,
+    ) -> Result<coordv1::Lease, kube_client::Error>
+    where
+        P: serde::Serialize + std::fmt::Debug,
+    {
+        tracing::debug!(?patch);
+        let params = kube_client::api::PatchParams {
+            field_manager: Some(self.field_manager.clone()),
+            // Force conflict resolution when using Server-side Apply (i.e., to
+            // acquire a lease). This is the recommended behavior for
+            // controllers.
+            force: matches!(patch, kube_client::api::Patch::Apply(_)),
+            ..Default::default()
+        };
+        self.api.patch(&*self.name, &params, patch).await
+    }
+
+    async fn get(api: Api, name: &str) -> Result<State, Error> {
+        let lease = api.get(name).await?;
+
+        let spec = match lease.spec {
+            Some(spec) => spec,
+            None => return Err(Error::MissingSpec),
+        };
```

Comment:
nit, TIOLI: could just be
```suggestion
        let spec = lease.spec.ok_or(Error::MissingSpec)?;
```